x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/mapper/SemanticTextInferenceResultFieldMapper.java

```diff
@@ -0,0 +1,241 @@
+        for (XContentParser.Token token = parser.nextToken(); token != XContentParser.Token.END_OBJECT; token = parser.nextToken()) {
+            if (token != XContentParser.Token.FIELD_NAME) {
+                throw new DocumentParsingException(parser.getTokenLocation(), "Expected a FIELD_NAME, got " + token);
+            }
+
+            parseFieldInferenceResults(context, mapperBuilderContext);
+        }
+    }
+
+    private static void parseFieldInferenceResults(
+        DocumentParserContext context,
+        MapperBuilderContext mapperBuilderContext) throws IOException {
+
+        String fieldName = context.parser().currentName();
+        Mapper mapper = context.getMapper(fieldName);
+        if (mapper == null || SemanticTextFieldMapper.CONTENT_TYPE.equals(mapper.typeName()) == false) {
+            throw new DocumentParsingException(context.parser().getTokenLocation(),
+                "Field [%s] is not registered as a %s field type".formatted(fieldName, SemanticTextFieldMapper.CONTENT_TYPE));
+        }
+
+        parseFieldInferenceResultsArray(context, mapperBuilderContext, fieldName);
+    }
+
+    private static void parseFieldInferenceResultsArray(
+        DocumentParserContext context,
+        MapperBuilderContext mapperBuilderContext,
+        String fieldName) throws IOException {
+
+        XContentParser parser = context.parser();
+        NestedObjectMapper nestedObjectMapper = createNestedObjectMapper(context, mapperBuilderContext, fieldName);
+        context.path().add(fieldName); // TODO: Need path manipulation?
+
+        try {
+            if (parser.nextToken() != XContentParser.Token.START_ARRAY) {
+                throw new DocumentParsingException(parser.getTokenLocation(), "Expected a START_ARRAY, got " + parser.currentToken());
+            }
+
+            for (XContentParser.Token token = parser.nextToken(); token != XContentParser.Token.END_ARRAY; token = parser.nextToken()) {
+                DocumentParserContext nestedContext = context
+                    .createChildContext(nestedObjectMapper)
+                    .createNestedContext(nestedObjectMapper);
+
+                if (token != XContentParser.Token.START_OBJECT) {
+                    throw new DocumentParsingException(parser.getTokenLocation(), "Expected a START_OBJECT, got " + parser.currentToken());
+                }
+
+                for (token = parser.nextToken(); token != XContentParser.Token.END_OBJECT; token = parser.nextToken()) {
+                    if (token != XContentParser.Token.FIELD_NAME) {
+                        throw new DocumentParsingException(parser.getTokenLocation(),
+                            "Expected a FIELD_NAME, got " + parser.currentToken());
+                    }
+
+                    String currentName = parser.currentName();
+                    context.path().add(currentName);
+                    try {
+                        // TODO: Test how this code handles extra/missing fields
```

Comment:
// TODO: Test how this code handles extra/missing fields
👍  We can probably rely on the NestedObjectMapper for that, but haven't checked it myself